### PR TITLE
 Add luaLibDir compiler option and in operator, fix template strings

### PR DESCRIPTION
**Pull request by zapp-brannigan-dota**
`luaLibDir` is a new compiler option that specifies the save location for `typescript_lualib.lua` relative to `outDir`.

Typescript template strings now cast all expressions to string.

### CODE REVIEW

File: src/Compiler.ts [hunk 1]
```diff
@@ -113,32 +132,41 @@ function removeInvalidOptionsFromArgv(commandLine: ReadonlyArray<string>, invali
...
     }
 }
 
 function executeCommandLine(args: ReadonlyArray<string>) {
     // Right now luaTarget, version and help are the only cli options, if more are added we should
     // add a more advanced custom parser
-    var argv = minimist(args, {
+    const argv = minimist(args, {
         string: ['l'],
         alias: { h: 'help', v: 'version', l: 'luaTarget' },
```
Comment (lolleko) [inline src/Compiler.ts]: Add CLI options for addHeader and lualiboutdir. Don't forget to update help text and readme.

File: src/Compiler.ts [hunk 2]
```diff
@@ -42,9 +47,23 @@ function compile(fileNames: string[], options: ts.CompilerOptions): void {
...
+    }
+
+    if (!options.luaLibDir) {
+        options.luaLibDir = options.outDir;
+    }
+    else {
+        options.luaLibDir = path.join(options.outDir, options.luaLibDir);
+    }
+    if (!fs.existsSync(options.luaLibDir)){
+        fs.mkdirSync(options.luaLibDir);
```
Comment (lolleko) [inline src/Compiler.ts]: Pretty sure mkdir sync is not recursive. In this case this will error on a nested path.

File: src/Compiler.ts [hunk 3]
```diff
@@ -42,9 +47,23 @@ function compile(fileNames: string[], options: ts.CompilerOptions): void {
...
 
+    if (!('addHeader' in options)) {
+        options.addHeader = true;
+    }
+
+    if (!options.luaLibDir) {
+        options.luaLibDir = options.outDir;
+    }
+    else {
+        options.luaLibDir = path.join(options.outDir, options.luaLibDir);
```
Comment (lolleko) [inline src/Compiler.ts]: should be path.resolve, just to be sure.

File: src/Compiler.ts [hunk 4]
```diff
@@ -11,7 +11,12 @@ import dedent = require("dedent")
 import { LuaTranspiler, TranspileError } from "./Transpiler";
 import { TSHelper as tsEx } from "./TSHelper";
 
-function compile(fileNames: string[], options: ts.CompilerOptions): void {
+interface CompilerOptions extends ts.CompilerOptions {
+    addHeader?: boolean;
+    luaLibDir?: string;
+}
```
Comment (lolleko) [inline src/Compiler.ts]: Add `luaTarget: string` aswell

File: src/Compiler.ts [hunk 5]
```diff
@@ -42,9 +47,23 @@ function compile(fileNames: string[], options: ts.CompilerOptions): void {
...
+        options.luaLibDir = path.join(options.outDir, options.luaLibDir);
+    }
+    if (!fs.existsSync(options.luaLibDir)){
+        fs.mkdirSync(options.luaLibDir);
+    }
+
     // Copy lualib to target dir
     // This isnt run in sync because copyFileSync wont report errors.
-    fs.copyFile(path.resolve(__dirname, "../dist/lualib/typescript.lua"), path.join(options.outDir, "typescript_lualib.lua"), (err: NodeJS.ErrnoException) => {
+    fs.copyFile(path.resolve(__dirname, "../dist/lualib/typescript.lua"), path.join(options.luaLibDir, "typescript_lualib.lua"), (err: NodeJS.ErrnoException) => {
```
Comment (lolleko) [inline src/Compiler.ts]: When you change options.luaLibDir the require call should be resolved, in order to point to that new path, https://github.com/Perryvw/TypescriptToLua/blob/master/src/Transpiler.ts#L36

File: src/Compiler.ts (hunk 2 above)
Comment (lolleko) [reply src/Compiler.ts]: Typescript has a function to recursively create files / directories. not sure If that function is exposed tho.

File: src/Compiler.ts [hunk 6]
```diff
@@ -113,32 +132,41 @@ function removeInvalidOptionsFromArgv(commandLine: ReadonlyArray<string>, invali
...
+                invalidErrorCount += 1;
+            }
+            else {
+                console.log(err.messageText);
+            }
         });
-        process.exit(1);
+        if (invalidErrorCount != commandLine.errors.length) {
+            process.exit(1);
+        }
```
Comment (Perryvw) [inline src/Compiler.ts]: all of `logError` looks pretty dodgy. Improve or add some comments that explain what/why this function does.

**Comment by Perryvw** [comment]
Not sure if we want to support luaLibDir in the first place, sounds like a lot of hacky stuff for little gain.

**Comment by Perryvw** [comment]
After some discussion we decided not to want the lualibdir stuff in the main branch. Like I said in the previous comment it seems like asking for trouble with limited pay-off. The rest of the pull request looks good and I'd be happy to merge if you remove the lualibdir part.

**Comment by zapp-brannigan-dota** [comment]
luaLibDir solves the problem of a directory structure like this:
```
src/
- game/
  - tsconfig.json
  - scripts/
    - vscripts/
      - ...
```
There is no additional resolving needed. It can only be used to specify the correct subfolder, nothing else.

File: src/Compiler.ts (hunk 5 above)
Comment (zapp-brannigan-dota) [reply src/Compiler.ts]: See comment below.

**Comment by lolleko** [comment]
Did you try setting rootDir to `./scripts/vscripts/`

**Comment by zapp-brannigan-dota** [comment]
rootDir solves the problem, so I removed luaLibDir. It was broken on Windows but is fixed now. I can't tell if it works on other platforms tho.

**Comment by zapp-brannigan-dota** [comment]
There should tests for compiling different types of skeleton projects. I might make them at some point but I've never done that so some examples would be very helpful.

**Pull request merged**